systemtests/auth_test.go

```diff
@@ -106,6 +107,175 @@ func (s *systemtestSuite) TestBuiltInAdminAuthorizationImmortality(c *C) {
+		c.Assert(len(body), Equals, 0)
+
+		// non-admins cannot access this endpoint
+		resp, _ = proxyDelete(c, userToken, endpoint+"/xxx")
+		c.Assert(resp.StatusCode, Equals, http.StatusForbidden)
+
+		// test 404
+		resp, _ = proxyDelete(c, adToken, endpoint+"/xxx")
+		c.Assert(resp.StatusCode, Equals, http.StatusNotFound)
+
+	})
+}
+
+// TestAuthorizationGetEndpoints tests authorization get endpoints (/authorizations, /authorizations/{ID})
+func (s *systemtestSuite) TestAuthorizationGetEndpoints(c *C) {
+	s.addUser(c, username)
+
+	runTest(func(ms *MockServer) {
+		userToken := loginAs(c, username, username)
+		endpoint := proxy.V1Prefix + "/authorizations"
+
+		// built-in admin user is always associated with a authz
+		c.Assert(len(s.getAuthorizations(c, adToken)) >= 1, Equals, true)
+
+		data := `{"PrincipalName":"` + username + `","local":true,"role":"ops","tenantName":"default"}`
+		// add authorization for `username` using admin token
+		authz := s.addAuthorization(c, data, adToken)
+		c.Assert(authz, DeepEquals, s.getAuthorization(c, authz.AuthzUUID, adToken))
+
+		// non-admins cannot access this endpoint
+		resp, _ := proxyGet(c, userToken, endpoint)
+		c.Assert(resp.StatusCode, Equals, http.StatusForbidden)
+
+		// add admin authz for `username`
+		data = `{"PrincipalName":"` + username + `","local":true,"role":"admin","tenantName":"default"}`
+		authz = s.addAuthorization(c, data, adToken)
+		c.Assert(authz.Role, Equals, "admin")
+		// tenantName gets ignored for admin authzs
+		c.Assert(authz.TenantName, Equals, "")
+		c.Assert(authz, DeepEquals, s.getAuthorization(c, authz.AuthzUUID, adToken))
+
+		// since `username` is an admin after the earlier update, he can access admin APIs now
+		data = `{"PrincipalName":"ops","local":true,"role":"admin","tenantName":""}`
+		s.addAuthorization(c, data, userToken)
+
+		// admin API accessed using user token
+		c.Assert(len(s.getAuthorizations(c, userToken)) >= 4, Equals, true)
+
+		// NOTE: there is no straight forward to demote a user from admin to ops.
+		// rather the same can be achieved using DELETE + ADD
+		s.deleteAuthorization(c, authz.AuthzUUID, userToken)
+		resp, _ = proxyGet(c, adToken, endpoint+"/"+authz.AuthzUUID)
+		c.Assert(resp.StatusCode, Equals, http.StatusNotFound)
+
+		// non-admins cannot access this endpoint
+		resp, _ = proxyGet(c, userToken, endpoint)
+		c.Assert(resp.StatusCode, Equals, http.StatusForbidden)
+
+		// non-admins cannot access `*/authzUUID` endpoint
+		resp, _ = proxyGet(c, userToken, endpoint+"/"+authz.AuthzUUID)
+		c.Assert(resp.StatusCode, Equals, http.StatusForbidden)
+
+		// delete all authorizations and try get
+		// NOTE: Built-in admin authorizations cannot be deleted
+		for _, authz := range s.getAuthorizations(c, adToken) {
+			resp, _ = proxyDelete(c, adToken, endpoint+"/"+authz.AuthzUUID)
+			c.Assert(resp.StatusCode == http.StatusNoContent || resp.StatusCode == http.StatusBadRequest, Equals, true)
```

Comment:
Can we use an if statement here and check for the proper return code for each item instead of allowing either one?